picard/util/pipe.py

```diff
@@ -0,0 +1,360 @@
+    MESSAGE = "Pipe is broken"
+
+
+class PipeErrorInvalidResponse(PipeError):
+    MESSAGE = "Invalid response from pipe"
+
+
+class PipeErrorWin(PipeError):
+    MESSAGE = "Windows API error"
+
+
+class PipeErrorNoPermission(PipeError):
+    MESSAGE = "No permissions for creating a pipe"
+
+
+class PipeErrorNoDestination(PipeError):
+    MESSAGE = "No available dirs to place a pipe"
+
+
+class AbstractPipe(metaclass=ABCMeta):
+    NO_RESPONSE_MESSAGE: str = "No response from FIFO"
+    MESSAGE_TO_IGNORE: str = "Ignore this message, just testing the pipe"
+    TIMEOUT_SECS: float = 1.5
+
+    @classmethod
+    @property
+    @abstractmethod
+    def PIPE_DIRS(cls):
+        raise NotImplementedError
+
+    def __init__(self, app_name: str, app_version: str, args=None, forced_path=None):
+        if args is None:
+            self._args = tuple()
+        else:
+            try:
+                self._args = tuple(args)
+            except TypeError as exc:
+                raise PipeErrorInvalidArgs(exc) from None
+
+        if not self._args:
+            self._args = (self.MESSAGE_TO_IGNORE,)
+
+        if not isinstance(app_name, str) or not isinstance(app_version, str):
+            raise PipeErrorInvalidAppData
+        elif IS_WIN:
+            app_version = app_version.replace(".", "-")
+
+        if forced_path:
+            self._paths = (forced_path,)
+        elif IS_WIN or os.getenv("HOME"):
+            self._paths = self.__generate_filenames(app_name, app_version)
+            self.path_was_forced = False
+        else:
+            self._paths = (NamedTemporaryFile(delete=False).name,)
+            self.path_was_forced = True
+            log.debug("Pipe path had to be mocked by a temporary file")
+        self.is_pipe_owner: bool = False
+
+        self.__thread_pool = concurrent.futures.ThreadPoolExecutor()
+
+    def _remove_temp_attributes(self):
+        del self._args
+        del self._paths
+
+    def __generate_filenames(self, app_name: str, app_version: str):
+        _pipe_names = []
+
+        for dir in self.PIPE_DIRS:
+            if dir:
+                _pipe_names.append(os.path.join(os.path.expanduser(dir),
+                                                        f"{app_name}_v{app_version}_pipe_file"))
+
+        if _pipe_names:
+            return _pipe_names
+
+        raise PipeErrorNoDestination
+
+    def _reader(self) -> str:
+        raise NotImplementedError()
+
+    def _sender(self, message) -> bool:
+        raise NotImplementedError()
+
+    def read_from_pipe(self, timeout_secs: Optional[float] = None) -> List[str]:
+        if timeout_secs is None:
+            timeout_secs = self.TIMEOUT_SECS
+
+        reader = self.__thread_pool.submit(self._reader)
+        out = []
+
+        try:
+            res = reader.result(timeout=timeout_secs)
+            if res:
+                res = res.split("\n")
+                for r in res:
+                    if res == self.MESSAGE_TO_IGNORE:
```

Comment:
I don't understand this line.

```python
if res == self.MESSAGE_TO_IGNORE:
```

Shouldn't it be `r` instead of `res`?

Reply:
It should.